Detect invalidated references statically (#2037)
onflow/cadence: docs/language/references.mdx

References are ephemeral, i.e. they cannot be [stored](accounts#account-storage).
Instead, consider [storing a capability and borrowing it](capability-based-access-control) when needed.

### Reference validity

Ephemeral references stay valid throughout the course of the program.

> However, **references to resources** can become invalid during a program if the underlying resource is moved

Reviewer: ```suggestion
However, **references to resources** can become invalid during the execution of a program, if the referenced resource is moved
```

or destroyed after taking the reference.

```cadence
let r <-create R()

// Take a reference to resource.
let ref = &r as &R

// Then transfer the resource into an account.
// This will invalidate all the references taken to the resource `r`.
account.save(<-r, to: /storage/r)

// Static error, since the referenced resource has been moved.
ref.id = 2
```

A reference is invalidated upon the first transfer of the underlying resource, regardless of the origin and the destination.

```cadence
let ref = &r as &R

> // Moving the resource to a different vairable would also invalidate the reference.

Reviewer: ```suggestion
// Moving a resource to a different variable invalidates all references of it.
```

Reviewer 2: I think usually we say "references to a value" rather than "references of a value", so maybe "invalidates all references to it" instead?

Author: ```suggestion
// Moving a resource to a different variable invalidates all references to it.
```

let r2 <- r

// Static error, since the referenced resource has been moved.
ref.id = 2
```

<Callout type="info">

> Invalidation of storage references are not statically caught, but happens at run-time.

Reviewer: ```suggestion
Invalidations of storage references are not statically caught, but only at run-time.
```
